### PR TITLE
Adds Swift Concurrency Availability to macOS 10.15, iOS 13, tvOS 13, and watchOS 6

### PR DESCRIPTION
**Pull request by PSchmiedmayer**
Adds Swift Concurrency Availability to macOS 10.15, iOS 13, tvOS 13, and watchOS 6

### Motivation:

This PR is related to #2000.
With the support for Swift Concurrency coming to macOS 10.15, iOS 13, tvOS 13, and watchOS 6 with Xcode 13.2, the Swift Concurrency features should be available when compiling for these targets using Xcode 13.2.

### Modifications:

Change the @available annotations to macOS 10.15, iOS 13, tvOS 13, and watchOS 6 and introduce a temporary code duplication as discussed in #2000.
Changes the source code of the NIOAsyncAwaitDemo to only work on Swift 5.2.2 (Xcode 13.2) on macOS 10.15 and newer and all other non-Darwin platforms.

### Result:

The Swift Concurrency features are marked with `@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)` and can therefore be used on macOS 10.15, iOS 13, tvOS 13, and watchOS 6.

### CODE REVIEW

File: Sources/NIOAsyncAwaitDemo/AsyncChannelIO.swift [hunk 1]
```diff
@@ -15,8 +15,8 @@
 import NIOCore
 import NIOHTTP1
 
-#if compiler(>=5.5) && canImport(_Concurrency)
-@available(macOS 12, iOS 15, tvOS 15, watchOS 8, *)
+#if canImport(_Concurrency) && compiler(>=5.5.2) || (!os(macOS) && compiler(>=5.5))
```
Comment (Lukasa) [inline Sources/NIOAsyncAwaitDemo/AsyncChannelIO.swift:18]: The `os` guards here are insufficient to cover the use-case (we'd need to check for all Apple platforms), but I don't think it matters. On non-Apple platforms the two function implementations are totally fine, so we can just drop the `||` here.

File: Sources/NIOAsyncAwaitDemo/main.swift [hunk 2]
```diff
@@ -16,9 +16,9 @@ import NIOPosix
 import NIOHTTP1
 import Dispatch
 
-#if compiler(>=5.5) && canImport(_Concurrency)
+#if canImport(_Concurrency) && compiler(>=5.5.2) || (!os(macOS) && compiler(>=5.5))
```
Comment (Lukasa) [inline Sources/NIOAsyncAwaitDemo/main.swift:19]: Similarly, we can drop the `||` here.

File: Sources/NIOCore/AsyncAwaitSupport+OldXcodes.swift [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if os(macOS) && compiler(>=5.5) && !compiler(>=5.5.2)
```
Comment (Lukasa) [inline Sources/NIOCore/AsyncAwaitSupport+OldXcodes.swift:15]: And we can drop the `os` guard here.

File: Sources/NIOAsyncAwaitDemo/AsyncChannelIO.swift (hunk 1 above)
Comment (PSchmiedmayer) [reply Sources/NIOAsyncAwaitDemo/AsyncChannelIO.swift]: Good point, I have updated the PR to incorporate your feedback 👍

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Review by Lukasa — APPROVED** [review]

**Comment by Lukasa** [comment]
Thanks for the patch!

**Comment by PSchmiedmayer** [comment]
Happy to help! Thanks for your fast response and your input @Lukasa & @weissi!

**Pull request merged**